atomic_cell: compare value last

bhalevy opened the pull request:
Currently, when two cells have the same write timestamp
and both are alive or expiring, we compare their value first,
before checking if either of them is expiring
and if both are expiring, comparing their expiration time
and ttl value to determine which of them will expire
later or was written later.

This was based on an early version of Cassandra.
However, the Cassandra implementation rightfully changed in
https://github.com/apache/cassandra/commit/e225c88a65f2e8091f8ea6212c291416674882a1 ([CASSANDRA-14592](https://issues.apache.org/jira/browse/CASSANDRA-14592)),
where the cell expiration is considered before the cell value.

To summarize, the motivation for this change is three fold:
1. Cassandra compatibility
2. Prevent an edge case where a null value is returned by select query when an expired cell has a larger value than a cell with later expiration.
3. A generalization of the above: value-based reconciliation may cause select query to return a mixture of upserts, if multiple upserts use the same timeastamp but have different expiration times.  If the cell value is considered before expiration, the select result may contain cells from different inserts, while reconciling based the expiration times will choose cells consistently from either upserts, as all cells in the respective upsert will carry the same expiration time.

Fixes #14182

Also, this series:
- updates dml documentation
- updates internal documentation
- updates and adds unit tests and cql pytest reproducing #14182

avikivity commented:
When did the change in the algorithm happen (commit)?

bhalevy commented:
> When did the change in the algorithm happen (commit)?

The problem was there early on, when f43836eb687542cbdf9d24ac162a80dd1f53a335 initially added handling of expirations, already evaluated after comparing the cell value.

avikivity commented:
> > When did the change in the algorithm happen (commit)?
> 
> The problem was there early on, when [f43836e](https://github.com/scylladb/scylladb/commit/f43836eb687542cbdf9d24ac162a80dd1f53a335) initially added handling of expirations, already evaluated after comparing the cell value.

Ok, so backport all the way to 0.10.

avikivity commented:
@tgrabiec please review. I don't have intuition about whether the old or new algorithm are more correct.

nyh commented:
A comment I wrote in the issue:
"If we change the order in our code, we may need to do something to avoid the above "oscillation" problem on mixed clusters during upgrade".
Do you think it's fine to not handle this issue?

nyh commented:
> also add a cql pytest to reproducing #14182.

We already have a test file, "test-timestamp.py", for tests of USING TIMESTAMP. Please add your test to that file - and feel free to rename it in your patch (I admit test_using_timestamp.py is a better name, and I don't know why this test ended up the only one having a minus instead of an underscore in its name...)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1721/

bhalevy commented:
> > also add a cql pytest to reproducing #14182.
> 
> We already have a test file, "test-timestamp.py", for tests of USING TIMESTAMP. Please add your test to that file - and feel free to rename it in your patch (I admit test_using_timestamp.py is a better name, and I don't know why this test ended up the only one having a minus instead of an underscore in its name...)

Done in [8cd6755](https://github.com/scylladb/scylladb/commit/8cd67559b7ceef7fc2cf923ff533ff5a5e22407f)

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1725/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1730/

vladzcloudius commented on mutation/atomic_cell.cc:
```diff
@@ -79,23 +79,20 @@ compare_atomic_cell_for_merge(atomic_cell_view left, atomic_cell_view right) {
         return left.is_live() ? std::strong_ordering::less : std::strong_ordering::greater;
```
I think that giving a live cell preference to not-live cell even if the not-live cell was written after the live cell is not obvious at all.

denesb replied:
We have no way to determine this. We use timestamps to determine order and we are in a tie. Our ordering broke down, and we just need *some* order (bonus if it is also intuitive).

vladzcloudius replied:
@denesb Don't we always store a coordinator time *together* with a "timestamp" (which can be overridden with `USING TIMESTAMP`)? 
Or we only have it if there if data is TTLed and we have an `expire_at`?

vladzcloudius replied:
@bhalevy is the code with this patch copies the latest C* logic 1-to-1?

bhalevy replied:
@vladzcloudius it doesn't _copy_ the C* logic but it is compatible with it, using a different implementation.
Also, it extend C*'s algorithm as we examine the cell remaining ttl, while C* doesn't so that.  They stop processing at the expiry time and if both cells expire at the same time, their value is compared.

vladzcloudius commented on test/boost/mutation_test.cc:
```diff
@@ -703,68 +700,85 @@ SEASTAR_TEST_CASE(test_cell_ordering) {
+        atomic_cell::make_live(*bytes_type, 1, bytes("value2")),
+        atomic_cell::make_live(*bytes_type, 1, bytes("value1"), expiry_1, ttl_1));
+
+    testlog.debug("Dead cells with same expiry are equal");
```
another unrelated change

vladzcloudius commented on test/boost/mutation_test.cc:
```diff
@@ -703,68 +700,85 @@ SEASTAR_TEST_CASE(test_cell_ordering) {
         atomic_cell::make_dead(1, expiry_1),
         atomic_cell::make_dead(1, expiry_1));
 
+    testlog.debug("Live cells are ordered before expiring cells, with empty value");
```
and another one

vladzcloudius commented on test/boost/mutation_test.cc:
```diff
@@ -703,68 +700,85 @@ SEASTAR_TEST_CASE(test_cell_ordering) {
     // Origin doesn't compare ttl (is it wise?)
     // But we do. See https://github.com/scylladb/scylla/issues/10156
     // and https://github.com/scylladb/scylla/issues/10173
+    testlog.debug("Expiring cells with higher ttl are ordered before expiring cells with smaller ttl and same expiry time");
```
and another one

vladzcloudius commented on test/boost/mutation_test.cc:
```diff
@@ -703,68 +700,85 @@ SEASTAR_TEST_CASE(test_cell_ordering) {
         atomic_cell::make_live(*bytes_type, 0, bytes("value1")),
         atomic_cell::make_live(*bytes_type, 0, bytes("value2")));
 
+    testlog.debug("Cells are ordered by value in lexicographical order if all else is equal");
```
and another one

vladzcloudius commented on test/boost/mutation_test.cc:
```diff
@@ -703,68 +700,85 @@ SEASTAR_TEST_CASE(test_cell_ordering) {
         atomic_cell::make_dead(1, expiry_1));
 
-    // Dead wins with expiring cell
+    testlog.debug("Dead wins with expiring cell");
```
and another one

vladzcloudius commented on test/boost/mutation_test.cc:
```diff
@@ -703,68 +700,85 @@ SEASTAR_TEST_CASE(test_cell_ordering) {
         atomic_cell::make_dead(2, expiry_1));
 
-    // ...then by expiry
+    testlog.debug("...then by expiry");
```
and another one

denesb commented on test/boost/mutation_test.cc:
```diff
@@ -703,68 +700,85 @@ SEASTAR_TEST_CASE(test_cell_ordering) {
+    testlog.debug("Live cells are ordered before expiring cells, regardless of their value");
+    assert_order(
+        atomic_cell::make_live(*bytes_type, 1, bytes("value2")),
+        atomic_cell::make_live(*bytes_type, 1, bytes("value1"), expiry_1, ttl_1));
```
Only expiring/dead cells have a write time. Live cells don't have one.

vladzcloudius replied:
I see. It kinda makes sense to me now too: they want to prefer the value that has the longest "expire_at" and the alive cell has an infinite value for this (formally).

bhalevy replied:
According to CASSANDRA-14592 they keep expiring cells because they'll become tombstones in the future and then they'll trump live cells.  If they kept the non-expiring cell (as it will survive longer), the expiration would be lost.  Their description of the issue indicates that this inconsistency depends on WHEN the comparison took place, as if it was made early enough, before the cell expired, then the live cell could have been picked (based on its value), but if the comparison took place later on, after the cell expired, then it would be considered as a tombstone and then it would prevail, so preferring expiring cells over live ones is there to be consistent with the rule the prefers tombstones over live cells.

vladzcloudius commented on mutation/atomic_cell.cc:
```diff
@@ -79,23 +79,20 @@ compare_atomic_cell_for_merge(atomic_cell_view left, atomic_cell_view right) {
                 return left.expiry() <=> right.expiry();
-            } else {
+            } else if (right.ttl() != left.ttl()) {
                 // prefer the cell that was written later,
```
This comment doesn't seem to reflect the need logic: the following comparison is going to order based on the TTL value - not the "expiry" (which would match the old comment).

You probably want to move this comment above line 87 and add a different comment here.

bhalevy replied:
The comment does refer to selection by ttl, although it's not obvious.
Having shorter ttl with same expiry means that the derived write time is later.
With that in mind, the cell becomes purgeable gc_grace_seconds after its write time, hence it will survive for longer (as a tombstone).

vladzcloudius reviewed (changes requested):
The code makes sense to me logically.
Beyond a few cosmetic changes I requested yesterday and today it looks good to me.

avikivity commented:
The patch  fixes the issue, but that's only because they are phrased the same way.

Issue: we should do X
Patch: do X

That's convenient, but it avoids the question of what's the right behavior, what's the user impact, and what should be written in the release notes.

bhalevy commented:
> The patch fixes the issue, but that's only because they are phrased the same way.
> 
> Issue: we should do X Patch: do X
> 
> That's convenient, but it avoids the question of what's the right behavior, what's the user impact, and what should be written in the release notes.

I edited the issue title and description to describe the user visible impact rather than the technical details.

nyh commented on test/cql-pytest/test_using_timestamp.py:
```diff
@@ -90,3 +90,56 @@ def test_key_writetime(cql, table1):
         cql.execute(f'SELECT ttl(k) FROM {table1}')
+
+# Reproducer for https://github.com/scylladb/scylladb/issues/14182
+def test_rewrite_using_same_timeout_and_ttl(scylla_only, cql, table1):
```
Wait, what?
I thought the main point of this patch was to behave like Cassandra always did in this specific case. So why a scylla_only test case?

bhalevy replied:
I wasn't sure at the time that we're indeed compatible.
Today I'm more confident that we are (with recent enough version of Cassandra).
I'll test that in the next version of this series.

nyh commented on test/cql-pytest/test_using_timestamp.py:
```diff
@@ -90,3 +90,56 @@ def test_key_writetime(cql, table1):
+            errors.append(f"Expected (k={k}, v={expected}) after immediate rewrite in iteration {i}, but got {res[0]}")
+
+        delay = 4
+        time.sleep(delay)
```
Sad you need such a long sleep. To be honest, I don't understand why you need to sleep at all, can't you read the data and its ttl() immediately and see this problem without needing to wait for the TTL?
This also relates to my question about what it is that we are **really** fixing in this PR, I'll write it as a separate comment.

bhalevy replied:
I'd like to reproduce the actual issue, rather than testing for indirect evidence.
But I'll see if it can be reproduced reliably using short timeouts.

nyh replied:
Sometimes after you understand an issue you understand that although the customer noticed the problem only after the expiration, if they used "SELECT ttl(...)" they would have seen it immediately after the second write, so you don't need to wait for the expiration. But I guess I'm ok with this if a couple of 1-second sleeps are involved, and not a dozen 4-second sleeps.

nyh commented on test/cql-pytest/test_using_timestamp.py:
```diff
@@ -90,3 +90,56 @@ def test_key_writetime(cql, table1):
+            errors.append(f"Expected (k={k}, v={expected}) after first expiration in iteration {i}, but got {res[0]}")
+
+        expire2 = t2 + ttl2
+        time.sleep(expire2 - time.time() + 2)
```
Please remember that typical cql-pytest tests take time measured in milliseconds, and think if there is no way to do this test without so many multi-second sleeps. I think no sleeps are really needed here (see above) but maybe I'm missing something?

bhalevy replied:
I wanted to reproduce the null value issue, and that requires the first insert to expire.

nyh reviewed (changes requested):
Although I understand this patch, I don't like the fact that it's **motivation** is ambiguous, and it makes it harder to review if the implementation lives up to the motivation. I can think of three related, but different motivations:

1. @vladzcloudius believes that this this patch fixes a real "error", and your commit message also claims it avoids a "weird corner case". What is this error / corner case, and without explaining what it is, how can I review that the code fixes it and the test reproduces it?
2. The fact that Cassandra's ordering is different from Scylla's was mentioned several times, and a possible motivation was to make them the same. But did we? The test is marked "scylla_only", suggesting it fails on Cassandra. But why?
3. Why I try to explain the motivation of this patch to myself, it's not about the TTL at all! Instead, it's about what to do when a user send different writes A and write B with the same timestamp - how can we try (if possible) to make sure either A or B gets written but not a mixture half-A-half-B? When writes have expiration-times, we are lucky:  In CQL, when you do a write, all of it will have the **same** expiration time. So all the cells in A have the same expiration time tA, and all the cells in B have the same expiration tB. If tA != tB (e.g., we have default-TTL and more than a second passed between the two writes), if we base the reconciliation decision on the expiration time, not cell values - we can consistently choose either A and B without mixing them!

I think the real motivation is, or should be, 3. Stating that this is the motivation isn't just nice for understanding, or for understanding how serious (or not) this issue is, it can also change the way you look at testing this issue. Benny, your test has a lot of sleeps waiting until the the data actually expires. Although it might make sense to have a slow test that exactly reproduces the same scenario that the user saw, I think you can actually reproduce the usefulness of this patch in one second (unfortunately, not less): Do a write A of two int cells with values 0, 1 with TTL 1000, sleep a whole second (unfortunately), then do a write B of the cells with values 1, 0 with TTL 1000. Now read. Before this patch, I believe you'll read a mixture of A and B (I think 0,0)). After this patch, you'll read one of A or B - not a mixture of the two. Also, unless I'm missing something, the same test should pass on Cassandra.

bhalevy commented:
In v2 ([ea9cb20](https://github.com/scylladb/scylladb/commit/ea9cb201936c506449ebc69b077442d6405cd551))
- rebased
- cleaned up changes to unit test (separating cosmetic changes from functional ones)
- added and update source documentation for `compare_atomic_cell_for_merge`
- Update user-facing doc: docs/cql/dml.rst with a new, `Update ordering` section

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1749/

bhalevy commented:
> Although I understand this patch, I don't like the fact that it's **motivation** is ambiguous, and it makes it harder to review if the implementation lives up to the motivation.

Good point.  Let's clear that out.

> I can think of three related, but different motivations:
> 
> 1. @vladzcloudius believes that this this patch fixes a real "error", and your commit message also claims it avoids a "weird corner case". What is this error / corner case, and without explaining what it is, how can I review that the code fixes it and the test reproduces it?

Sure.  I'll describe the case more clearly.

> 2. The fact that Cassandra's ordering is different from Scylla's was mentioned several times, and a possible motivation was to make them the same. But did we? The test is marked "scylla_only", suggesting it fails on Cassandra. But why?

Today, after reviewing both implementation, I convinced myself the implementation are now compatible and the scylla_only marker can be removed (will do in next version of the series)

> 3. Why I try to explain the motivation of this patch to myself, it's not about the TTL at all! Instead, it's about what to do when a user send different writes A and write B with the same timestamp - how can we try (if possible) to make sure either A or B gets written but not a mixture half-A-half-B? When writes have expiration-times, we are lucky:  In CQL, when you do a write, all of it will have the **same** expiration time. So all the cells in A have the same expiration time tA, and all the cells in B have the same expiration tB. If tA != tB (e.g., we have default-TTL and more than a second passed between the two writes), if we base the reconciliation decision on the expiration time, not cell values - we can consistently choose either A and B without mixing them! If we're lucky, we shouldn't waste this luck!

That's a good point and another motivator for this change (that I didn't think about).

> I think the real motivation is, or should be, 3. Stating that this is the motivation isn't just nice for understanding, or for understanding how serious (or not) this issue is, it can also change the way you look at testing this issue. Benny, your test has a lot of sleeps waiting until the the data actually expires. Although it might make sense to have a slow test that exactly reproduces the same scenario that the user saw, I think you can actually reproduce the usefulness of this patch in one second (unfortunately, not less) - without waiting for anything to expire: Do a write A of two int cells with values 0, 1 with TTL 1000, sleep a whole second (unfortunately), then do a write B of the cells with values 1, 0 with TTL 1000. Now read. Before this patch, I believe you'll read a mixture of A and B (I think 0,0)). After this patch, you'll read one of A or B - not a mixture of the two. Also, unless I'm missing something, the same test should pass on Cassandra.

I'll try that.

bhalevy commented:
> When did the change in the algorithm happen (commit)?

@avikivity our original implementation was based on a respectively old version of Cassandra.
They changed their algorithm in https://github.com/apache/cassandra/commit/e225c88a65f2e8091f8ea6212c291416674882a1 ([CASSANDRA-14592](https://issues.apache.org/jira/browse/CASSANDRA-14592)), but we didn't update our implementation correspondingly.

bhalevy commented:
In v3 ([3214dbf](https://github.com/scylladb/scylladb/commit/3214dbf17088c830ccd5ec6000a970e2222088c3)):
- Improved user-facing docs
- Improved annotations in git commits
- cql-pytest
  - reduce timeouts
  - removed `scylla_only` marker
  - added test case for updating multiple cells
- updated description of this PR to clarify the motivation for the change.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1752/

bhalevy commented:
In [6c98ed4](https://github.com/scylladb/scylladb/commit/6c98ed4cb73638878d4422a197f62e89292aac0d): changed git commit log to `Refs` https://github.com/scylladb/scylladb/issues/14063 rather than `Fixes`.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1753/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1754/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1756/

nyh commented:
Some comments about the new cover letter:

> Currently, when two cells have the same write timestamp and both are alive or expiring, we compare their value first, before checking if either of them is expiring and if both are expiring, comparing their expiration time and ttl value to determine which of them will expire later or was written later.

I think the phrase "was written later" is misleading. As I understand your code, you don't check which write was done "later". All we know is that in CQL as it is today, a single write statement (INSERT or UPDATE) will always set every touched cell to the same "expiration time", so we can consistently choose between two separate writes in all cells. We don't know if what we chose is "later" or "earlier".

If you really wanted to see which write was "later" I think you should calculate `expire - ttl`, and use that as an estimate (one second resolution) for the item's write time. But I don't see your code doing that.

> 
> This was based on an early version of Cassandra. However, the Cassandra implementation rightfully changed in [apache/cassandra@e225c88](https://github.com/apache/cassandra/commit/e225c88a65f2e8091f8ea6212c291416674882a1) ([CASSANDRA-14592](https://issues.apache.org/jira/browse/CASSANDRA-14592)), where the cell expiration is considered before the cell value.

I'm confused - are you sure that this Cassandra issue was about the same situation that we're fixing here, and not yet another corner case?

nyh commented on test/boost/mutation_test.cc:
```diff
@@ -710,6 +710,11 @@ SEASTAR_TEST_CASE(test_cell_ordering) {
         atomic_cell::make_live(*bytes_type, 1, bytes("value")),
         atomic_cell::make_live(*bytes_type, 1, bytes("value"), expiry_1, ttl_1));
 
+    testlog.debug("Live cells are ordered before expiring cells, regardless of their value");
```
"live" cells does not contrast with "expiring" cells. It seems that in your example, both cells are live - just one has a ttl and one doesn't.

bhalevy replied:
ok, will clarify message.

bhalevy replied:
Oops, I changed similar comments but missed this one...
Will fix.

bhalevy replied:
fixed in next version (f298a9e485)

nyh commented on test/boost/mutation_test.cc:
```diff
@@ -748,6 +753,16 @@ SEASTAR_TEST_CASE(test_cell_ordering) {
         atomic_cell::make_live(*bytes_type, 1, bytes(), expiry_1, ttl_1),
         atomic_cell::make_live(*bytes_type, 1, bytes(), expiry_2, ttl_1));
 
+    testlog.debug("...then by ttl (in reverse)");
```
I don't like these C++ tests because it's not possible to run them against Cassandra and check if we just invented our own rules, or Cassandra also follows the same rules.

bhalevy replied:
Maybe we should extend the cql pytest to do that so it can be tested against the reference.
Keep in mind just that the ttl rule is scylla-only, as cassandra stops at the expiry level and if the cells expire at the same time, the one with the higher value (lexicographically) is selected, regardless of their remaining ttl.

nyh replied:
So we should ask ourselves why we need a different rule. Although I do agree checking the TTL as a penultimate check does make sense - for the same reason that all the cells in a single write will have the same "ttl", not just same expiration. In any case, it's good to have tests that demonstrate how we differ from Cassandra - I don't think you said what you said here (that we won't be identical to Cassandra in this case) in any of the commit messages, for example.

bhalevy replied:
It was mentioned, though quite timidly in a57c087c897f60c331d8c9c3abcb125aef9250a2
but it can be expressed more visibly and explicitly tested.

bhalevy replied:
In v4, the deviation is documented more clearly in a C++ comment
and a scylla_only pytest tests this

nyh commented on mutation/atomic_cell.cc:
```diff
@@ -66,24 +66,27 @@ atomic_cell::atomic_cell(const abstract_type& type, atomic_cell_view other)
-            // prefer expiring cells.
             return left.is_live_and_has_ttl() ? std::strong_ordering::greater : std::strong_ordering::less;
         }
+        // If both are expiring, we do not want to consult the value bytes if we can avoid it, as like with C-14592
```
I didn't understand why this is like CASSANDRA-14592 (please don't shorthand Cassandra's bug-number convention). But anyway, since we have our own issue about this now, probably should mention that?

bhalevy replied:
The comment was taken from them btw (including the shorthand).
I think using our issue number would be better indeed. thanks.
will change.

nyh commented on docs/cql/dml.rst:
```diff
@@ -601,7 +601,7 @@ by the same node are unique. Timestamps assigned at different
 With a steadily high write rate timestamp collision
 is not unlikely. If it happens, i.e. two ``INSERTS`` have the same
-timestamp, the lexicographically bigger value prevails:
+timestamp, a tie-breaking algorithm determines which if the inserts prevails (see :ref:`update ordering <update-ordering>`).
```
This phrasing (also before your change, but also after) seems to suggest than one of the two INSERTS will prevail, while in reality, under some conditions, the result can be a weird mix of the two (with each column getting the maximum value from the two inserts).

bhalevy replied:
I agree that the text is quite vague but taking it simply, it does imply that the INSERT is considered in whole, rather than selecting which cell will prevail, individually.

We can add a note about that in the update-ordering section.

bhalevy replied:
Also, not that we can't guarantee that "insert atomicity" if neither of the inserts have a TTL.
In that case, the cells values are used for cell selection, independently, and the result could very well be a mixture of the inserts (since a. we compare the cells independently and not the whole row-mutation, and b. we don't keep a write time on non-expiring cells that could use to break the tie more elegantly)

nyh replied:
Exactly, which is why I'm saying that it's not a great idea to have wording that suggests we can always choose "which insert prevails".

nyh commented on test/cql-pytest/test_using_timestamp.py:
```diff
@@ -90,3 +90,79 @@ def test_key_writetime(cql, table1):
     with pytest.raises(InvalidRequest, match='PRIMARY KEY part k'):
         cql.execute(f'SELECT ttl(k) FROM {table1}')
+
+# Reproducer for https://github.com/scylladb/scylladb/issues/14182
```
If you already bothered to write a paragraph-long explanation of this test (and the second test) in the commit message, it will be great to copy it here. So in a few years if this test fails, the reader won't need to start guessing what this test was supposed to check.

Nitpick: It's great that you mentioned the issue that this reproduces, but you can also write "#14182" instead of the long github link.

bhalevy replied:
I dislike the `#14182` as it implies the scylladb.git repo.
This breaks when this patch is merged into the scylla-enterprise repo.

nyh commented on test/cql-pytest/test_using_timestamp.py:
```diff
@@ -90,3 +90,79 @@ def test_key_writetime(cql, table1):
+        t1 = time.time()
+        while t1 - int(t1) >= 0.9:
+            time.sleep(1 - (t1 - int(t1)))
+            t1 = time.time()
```
I think this is risky. On a very slow machine it can do weird things. Why is this even necessary?

bhalevy replied:
ttl is quantized in full second resolution.
If the two consecutive inserts cross the full second boundary they will not have the same expiry anymore and this makes the test flaky (I actually reproduced that in debug mode with a millisecond to go before the full-second mark). 10 milliseconds margin proved to be reliable enough on my local machine  in debug mode.

bhalevy commented:
> Some comments about the new cover letter:
> 
> > Currently, when two cells have the same write timestamp and both are alive or expiring, we compare their value first, before checking if either of them is expiring and if both are expiring, comparing their expiration time and ttl value to determine which of them will expire later or was written later.
> 
> I think the phrase "was written later" is misleading. As I understand your code, you don't check which write was done "later". All we know is that in CQL as it is today, a single write statement (INSERT or UPDATE) will always set every touched cell to the same "expiration time", so we can consistently choose between two separate writes in all cells. We don't know if what we chose is "later" or "earlier".
> 
> If you really wanted to see which write was "later" I think you should calculate `expire - ttl`, and use that as an estimate (one second resolution) for the item's write time. But I don't see your code doing that.

The existing code already does this, though it might not be obvious.
After comparing the `expiry()` time point, when it's equal on both cells, we compare the `ttl()` value and choose the cell with the smaller `ttl`.  As the comment there says, a smaller `ttl` for the same `expiry` means a later write time.
> 
> > This was based on an early version of Cassandra. However, the Cassandra implementation rightfully changed in [apache/cassandra@e225c88](https://github.com/apache/cassandra/commit/e225c88a65f2e8091f8ea6212c291416674882a1) ([CASSANDRA-14592](https://issues.apache.org/jira/browse/CASSANDRA-14592)), where the cell expiration is considered before the cell value.
> 
> I'm confused - are you sure that this Cassandra issue was about the same situation that we're fixing here, and not yet another corner case?

The Cassandra change does not mention this issue, but practically, it prevents its.
The motivation for https://issues.apache.org/jira/browse/CASSANDRA-14592 as I understand is more about considering the cell expiration so that expiring cells are preferred over live cells with the same write timestamp, before the expiring cell becomes a tombstone, in which case it will be preferred over the live cell, based on an earlier term in the algorithm.

nyh commented on test/cql-pytest/test_using_timestamp.py:
```diff
@@ -90,3 +90,79 @@ def test_key_writetime(cql, table1):
+            t1 = time.time()
+
+        cql.execute(f"INSERT INTO {table} (k, v) VALUES ({k}, {v1}) USING TIMESTAMP {ts} AND TTL {ttl1}")
+        # rewriting right away, so that both cells will carry the same expiry and ttl should return the larger value
```
I wouldn't test this case at all. The purpose of the test was supposed to be to check the case where the expiry is NOT the same, and you can achieve this by setting different TTL and not by weird sleeps.
I guess you can also do *this* check but I would make it a separate test (if it exists at all).

bhalevy replied:
Separated into a different test case in v4

nyh commented on test/cql-pytest/test_using_timestamp.py:
```diff
@@ -90,3 +90,79 @@ def test_key_writetime(cql, table1):
+        cql.execute(rewrite)
+        select = f"SELECT k, v FROM {table} WHERE k = {k}"
+        res = list(cql.execute(select))
+        assert len(res) == 1, f"Expected a single row, but got {res}"
```
General comment about pytest: `assert len(res) == 1` is enough, you don't need to invent a different failure message. That's one of the most beautiful features of pytest: It instruments "assert()" calls to give you a lot of extra information when it fails. In this case it will tell you exactly which assert call fails (the entire call tree), and tell you what was `len(res)` and even what was `res` when this failed. It's beautiful.

bhalevy replied:
yeah, I'm too familiar with pytest assertion failure messages.
The only problem with them is that they abbreviate long values, while the hand-written error message provides the full `res` list.

nyh replied:
Almost two years ago, pytest added the "-vv" option (the v option twice) that will avoid this truncation of long values.
Note that even when long lists are involved, pytest many times managest to print the right part. For example, if you do "assert l1 == l2", it doesn't prints specifically which part of l1 and l2 differ, not a hundred items which are equal. In our C++ tests and dtest we jumped through hoops to add these features, but pytest gives it to use for free - you just need to use "assert". This pytest feature is called "assertion rewriting" and is perhaps my favorite part of pytest.

Check out:
https://www.pythoninsight.com/2018/01/assertion-rewriting-in-pytest-part-1/
https://www.pythoninsight.com/2018/02/assertion-rewriting-in-pytest-part-4-the-implementation/

bhalevy replied:
got rid of the customized error messages in v4

nyh commented on test/cql-pytest/test_using_timestamp.py:
```diff
@@ -90,3 +90,79 @@ def test_key_writetime(cql, table1):
+    # sync with wall-clock on exact second so that expiration won't cross the whole-second boundary
+    while t1 - int(t1) >= 0.9:
+        time.sleep(1 - (t1 - int(t1)))
+        t1 = time.time()
```
This loop isn't needed here at all. Since you do a sleep(1) below between the first and second writesd, their expiration time will be **at least** one second difference. It doesn't matter if the first write is close to the beginning of a second or its end.

bhalevy replied:
Hmm, it makes sense in this test case.

nyh commented on test/cql-pytest/test_using_timestamp.py:
```diff
@@ -90,3 +90,79 @@ def test_key_writetime(cql, table1):
+    # if reconciliation is done by value, the result will be a mix of the two writes
+    # while if reconciliation is based first on the expiration time, the second write should prevail.
+    cql.execute(f"INSERT INTO {table} (k, v, w) VALUES ({k}, {values[0]['v']}, {values[0]['w']}) USING TIMESTAMP {ts} AND TTL {ttl}")
+    time.sleep(1)
```
Note: Instead of this sleep(1), you could have also just used a different TTL for the two writes - e.g., 1000 and 2000. They will have very different expiration times (and because you look at the expiration time first, before ttl, it doesn't matter that the ttls differ).

bhalevy replied:
true, will change

nyh commented on test/cql-pytest/test_using_timestamp.py:
```diff
@@ -90,3 +90,79 @@ def test_key_writetime(cql, table1):
+
+    res = list(cql.execute(f"SELECT * FROM {table} WHERE k = {k}"))
+    assert len(res) == 1, f"Expected a single row, but got {res}"
+    assert res[0].k == k and res[0].v == values[1]['v'] and res[0].w == values[1]['w'], \
```
As before, all these messages are unnecessary, and just hide the fact that pytest is fantastic and an order of magnitude easier to write and debug tests than Boost tests ;-)

bhalevy replied:
cleaned up in v4

nyh reviewed (approved):
Looks good. I still have comments and requests, but their are relatively minor and not critical.

bhalevy commented on test/boost/mutation_test.cc:
```diff
@@ -687,14 +687,11 @@ SEASTAR_TEST_CASE(test_cell_ordering) {
-            testlog.trace("Expected {} < {}", second, first);
-            abort();
-        }
+        testlog.trace("Expected {} < {}", first, second);
```
The trace logs are needed since we can't use `BOOST_REQUIRE_LT(compare_atomic_cell_for_merge(first, second), 0)` because boost doesn't know how to print `std::strong_ordering` and the boost printout on failure is too cryptic for `BOOST_REQUIRE(compare_atomic_cell_for_merge(first, second) < 0)`, making the trace logging useful when trying to figure out what exactly happened (the comparison can fail 2 ways: when comparing left, right and when comparing right, left).

bhalevy commented:
In v4 ([a749f0c](https://github.com/scylladb/scylladb/commit/a749f0c686b55723e0dbc99840d6bba28b6db1b5)):
- atomic_cell: compare_atomic_cell_for_merge: update and add documentation
  - Improved comments
  - Added note about extending C*'s algorithm when comparing ttl
- docs: dml: update tie breaking algorithm description when using timestamp
  - fixed a couple typos
  - added note to "Update ordering" section about atomicity of multi-column upserts (and referral to LWT)
- cql-pytest: test_using_timestamp: add tests for rewrites using same timestamp
  - cleaned up and simplified tests
  - separated out `test_rewrite_different_values_using_same_timestamp`
  - and scylla_only `test_rewrite_different_values_using_same_timestamp_and_expiration`

nyh commented on mutation/atomic_cell.cc:
```diff
@@ -66,32 +66,43 @@ atomic_cell::atomic_cell(const abstract_type& type, atomic_cell_view other)
+                //
+                // Note that this is an extension to Cassandra's algorithm
+                // which stops at the expiration time, and if equal,
+                // move forward to compare the cell values.
```
Good comment.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1765/

bhalevy commented:
In [f298a9e](https://github.com/scylladb/scylladb/commit/f298a9e4854235bd01d1a775a0caf40e029e9df9):
- fixed debug message in `test_cell_ordering`

nyh reviewed (approved):
Benny, let's make a deal - I'll agree to commit this PR and its tests if you agree to fix them if any of them is discovered to be flaky in the future ;-)

bhalevy commented:
> Benny, let's make a deal - I'll agree to commit this PR and its tests if you agree to fix them if any of them is discovered to be flaky in the future ;-)

It's a deal :)

bhalevy commented on docs/cql/dml.rst:
```diff
@@ -759,6 +759,37 @@ parameters:
+
+If both cells are expiring, the one with the latest expiration time wins;
+and if they have the same expiration time (in whole second resolution),
+their write time is derived from the expiration time less the remaining time-to-live
```
sigh, we shouldn't get into the gory technical details, but you're technically correct.  It's the ttl used when applying the cell.

bhalevy commented on docs/cql/dml.rst:
```diff
@@ -759,6 +759,37 @@ parameters:
+the cell with the lexicographically bigger value prevails.
+
+Note that when a TTL is used and multiple columns are INSERTed or UPDATEed using the same timestamp,
+SELECTing those columns will return cells picked from only one of those upserts, that has the latest expiration time,
```
so should we mention this at all?
since we have so many gotchas to this "feature" it's very hard for the application to make use of it anyhow.

bhalevy commented on test/cql-pytest/test_using_timestamp.py:
```diff
@@ -18,10 +18,20 @@
+    while t - int(t) >= 1 - millis / 1000:
+        time.sleep(1 - (t - int(t)))
+        t = time.time()
+    return t
```
The test I added also tests the "by value" case where there are 2 inserts that expire at the same time.
To achieve that they need to be written at the same whole-second period.  If one is inserted just before the 1-second mark and the other right after it, they won't have the same expiration time anymore.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1768/

bhalevy commented:
In v5 ([cc63305](https://github.com/scylladb/scylladb/commit/cc63305d39597534d337de476464ea0a9bb08e93)):
- updated docs/cql/dml.rst
  - clarified text about expiration time and about atomicity of updates with that respect, @annastuchlik please review docs patch.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1775/

vladzcloudius commented on docs/cql/dml.rst:
```diff
@@ -601,7 +601,7 @@ by the same node are unique. Timestamps assigned at different
 With a steadily high write rate timestamp collision
 is not unlikely. If it happens, i.e. two ``INSERTS`` have the same
-timestamp, the lexicographically bigger value prevails:
+timestamp, a tie-breaking algorithm determines which of the inserts prevails (see :ref:`update ordering <update-ordering>`).
```
System behavior **must** be documented. Period! There is nothing worse and "undefined behavior".
What users should or should not do - leave to users to decide.
If you are personally so against some particular practices you are welcome to write a "Best Practices" article and reference it from the document page mentioned above. 
I suggest we continue this discussion on the PR that adds this article.

bhalevy replied:
I can drop the patch that updates the documentation from this series if it's still controversial.
I think it's okay to document our behavior, but we should still be careful to be clear about what we want and can guarantee v.s. what we consider as implementation details that are subject to change, and also timing-sensitive so may produce different result due to server time-skew and/or low resolution of ttl - and so we advise the users not to step into those grey areas.

bhalevy replied:
Added disclaimer in v6

tgrabiec replied:
What about 

> What if someone relied on "the lexicographically bigger value prevails"?

The old behavior was documented, so a user could have "decided" to rely on it.

bhalevy replied:
I see your point, but what exactly do you think we should do?

nyh replied:
I agree with @tgrabiec, I think we can document that if a user does two writes with the same timestamp, the result MAY be a mixed version of the two writes - without explain exactly which extenuating circumstances (such as TTL and conflicting writes more than one second apart) may ensure that the conflict will result in one of the two writes but not n a mix.

I'd love to have a developer-oriented documentation (in docs/dev/) for the conflict resolution details, but I don't think users should make use of these details (and if they did, now with this PR, they will be frustrated because the details changed).

bhalevy commented on mutation/atomic_cell.cc:
```diff
@@ -79,23 +79,20 @@ compare_atomic_cell_for_merge(atomic_cell_view left, atomic_cell_view right) {
         return left.is_live() ? std::strong_ordering::less : std::strong_ordering::greater;
     }
     if (left.is_live()) {
```
It deviates from the current C* implementation.

bhalevy replied:
commitlog message improved in v6

bhalevy commented on docs/cql/dml.rst:
```diff
@@ -748,8 +748,8 @@ parameters:
-  Note that with a steadily high write rate, timestamp collision is not unlikely. If it happens, e.g. two INSERTS
-  have the same timestamp, conflicting cell values are compared and the cells with the lexicographically bigger value prevail.
+  Note that with a steadily high write rate, timestamp collision is not unlikely. If it happens, e.g. two INSERTS or UPDATES
+  have the same timestamp, a tie-breaking algorithm determines which of the upserts prevails (see :ref:`update ordering <update-ordering>` for more information):
```
Good point.  I'll phrase it closer to the original phrase that referred to the individual cells as the unit of comparison.

bhalevy replied:
improved in v6

bhalevy commented on mutation/mutation_partition.cc:
```diff
@@ -1108,20 +1108,34 @@ operator<<(std::ostream& os, const mutation_partition::printer& p) {
 
+// Note: the ordering algorithm for rows is the same as for cells,
+// except that there is no cell value to break a tie in case all other attributes are equal.
+// See compare_atomic_cell_for_merge.
```
yeah, let's leave it as a possible follow up

annastuchlik commented on docs/cql/dml.rst:
```diff
@@ -759,6 +759,51 @@ parameters:
+
+DISCLAIMER: The algorithm may be subject to change.
+The actual results depend on real-time differences between the cluster servers'
+clocks and is limited by the 1-second resolution of the TTL value.
```
```suggestion
clocks and are limited by the 1-second resolution of the TTL value.
```

annastuchlik commented on docs/cql/dml.rst:
```diff
@@ -759,6 +759,51 @@ parameters:
+the table's default TTL, or explicit, ``USING TTL``) are due to expire
+TTL seconds after the time they were written (as determined by the coordinator,
+and rounded to 1 second resolution). That time is the cell's expiration time.
+When cells expire they become tombstones, shadowing any data written with a write timestamp
```
```suggestion
When cells expire, they become tombstones, shadowing any data written with a write timestamp
```

annastuchlik commented on docs/cql/dml.rst:
```diff
@@ -759,6 +759,51 @@ parameters:
+
+Note that when multiple columns are INSERTed or UPDATEed using the same timestamp,
+SELECTing those columns might return a result that mixes cells from either upsert
+if both upserts have no expiration time, or their expiration time and ttl are the
```
```suggestion
This may happen when both upserts have no expiration time or their expiration time and TTL are the
```

annastuchlik commented on docs/cql/dml.rst:
```diff
@@ -759,6 +759,51 @@ parameters:
+Note that when multiple columns are INSERTed or UPDATEed using the same timestamp,
+SELECTing those columns might return a result that mixes cells from either upsert
+if both upserts have no expiration time, or their expiration time and ttl are the
+same, in whole second resolution; as cell selection would be based on the cell values
```
```suggestion
same (in whole second resolution). In such a case, cell selection will be based on the cell values
```

raphaelsc commented on mutation/mutation_partition.cc:
```diff
@@ -1108,20 +1108,34 @@ operator<<(std::ostream& os, const mutation_partition::printer& p) {
+                // The cell write time is derived by (expiry - ttl).
+                // Prefer row that was written later (and has a smaller ttl).
+                return left.ttl() < right.ttl() ? 1 : -1;
+            }
```
@bhalevy does this merit a backport?

raphaelsc replied:
the reason I asked this is because  I didn't see an issue opened specifically for row marker missing the logic, but perhaps there's no need for one at this patch will be included in the backport anyway.

bhalevy commented:
In v7 ([9f4e84c](https://github.com/scylladb/scylladb/commit/9f4e84c67b80ef072e678797c8e89e7477ef5d5b)):
- updated dml.doc based on the review comments and https://github.com/scylladb/scylladb/pull/14183#issuecomment-1589293308
 - added docs/dev/timestamp-conflict-resolution.md to document the details of the algorithm
 - dml test deals with the more generic topoics and explain the higher-level properties of the conflict resolution algorithm
 - s/tie-breaking/conflict resolution/ when it is referenced in other sections.

@avikivity one more issue I discussed with @tgrabiec is the roll-out of the change.
Ideally, we may want to add a config option (false for existing clusters and true for new ones)
for backward compatibility, plus cluster feature to enable the new algorithm on all nodes in the cluster only after rolling upgrade completes, since the conflict resolution should be consistent on all nodes.
OTOH, this case is so rare I'm not sure it merits the extra safety and the alight performance penalty it might carry (that is very small but happens on frequently-called code)

annastuchlik commented on docs/cql/dml.rst:
```diff
@@ -759,6 +759,57 @@ parameters:
+However, it is possible that multiple such cells will carry the same ``TIMESTAMP``.
+There could be several reasons for ``TIMESTAMP`` collision:
+
+1. Benign collision can be caused by "replay" of a mutation, e.g. due to client retry, or due to internal processes.
```
```suggestion
* Benign collision can be caused by "replay" of a mutation, e.g., due to client retry, or due to internal processes.
```

bhalevy replied:
fixed in v8

annastuchlik commented on docs/cql/dml.rst:
```diff
@@ -759,6 +759,57 @@ parameters:
+There could be several reasons for ``TIMESTAMP`` collision:
+
+1. Benign collision can be caused by "replay" of a mutation, e.g. due to client retry, or due to internal processes.
+In such case the cells are equivalent and any of them can be selected, arbitrarily.
```
```suggestion
In such cases, the cells are equivalent, and any of them can be selected arbitrarily.
```

bhalevy replied:
fixed in v8

annastuchlik commented on docs/cql/dml.rst:
```diff
@@ -759,6 +759,57 @@ parameters:
+1. Benign collision can be caused by "replay" of a mutation, e.g. due to client retry, or due to internal processes.
+In such case the cells are equivalent and any of them can be selected, arbitrarily.
+
+2. ``TIMESTAMP`` collisions might be normally caused by parallel queries that are served
```
```suggestion
* ``TIMESTAMP`` collisions might be normally caused by parallel queries that are served
```

bhalevy replied:
fixed in v8

annastuchlik commented on docs/cql/dml.rst:
```diff
@@ -759,6 +759,57 @@ parameters:
+by different coordinator nodes. The coordinators might calculate the same write ``TIMESTAMP``
+based on their local time in microseconds.
+
+1. Collisions might also happen with user-provided timestamps, if the application does not guarantee
```
```suggestion
* Collisions might also happen with user-provided timestamps if the application does not guarantee
```

bhalevy replied:
fixed in v8

bhalevy commented:
In v8 ([6dd2e15](https://github.com/scylladb/scylladb/commit/6dd2e155356269af9c22d1d9acb7fbb6b2ecbf5f)):
- Fixed @annastuchlik doc review comments (thanks for the prompt review!)

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1821/

nyh commented:
> CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1821/

It seems that the Alternator test hung, but I can't figure out where I got completely lost in the log files which keep moving around :-( I'll restart the CI.

@tgrabiec you have a "request changes" mark so please re-review this version. @annastuchlik maybe you also want to review, you had some comments on the last version.

avikivity commented on test/boost/mutation_test.cc:
```diff
@@ -703,68 +700,75 @@ SEASTAR_TEST_CASE(test_cell_ordering) {
         atomic_cell::make_live(*bytes_type, 1, bytes("value1")));
 
-    // ..then by value
+    testlog.debug("...then by value");
```
Will this cause a huge expansion in the logs, or is the default level for tests INFO?

(or the repetition count is low enough)?

bhalevy replied:
The default level is INFO and the repetition count is low (single digit)

avikivity commented on mutation/mutation_partition.cc:
```diff
@@ -1108,20 +1108,34 @@ operator<<(std::ostream& os, const mutation_partition::printer& p) {
+                // The cell write time is derived by (expiry - ttl).
+                // Prefer row that was written later (and has a smaller ttl).
+                return left.ttl() < right.ttl() ? 1 : -1;
+            }
```
This is a copy of the atomic_cell algorithm, yes? We should consolidate it into a template (but not in this series).

avikivity replied:
Also change to std::strong_ordering.

bhalevy replied:
a template or maybe a function that gets the individual attributes and a value reference that can be set to zero by row markers.

avikivity replied:
Yes

avikivity reviewed (approved):
Looks good, asked a minor question earlier.

@tgrabiec should have the final say + merge.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1906/

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1906/

That weird.  The failure, https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1906/artifact/testlog/x86_64/release/cql-pytest.test_using_timestamp.1.log:
```
    def test_key_writetime(cql, table1):
>       with pytest.raises(InvalidRequest, match='PRIMARY KEY part k'):
E       AssertionError: Regex pattern 'PRIMARY KEY part k' does not match 'Error from server: code=2200 [Invalid query] message="WRITETIME is not legal on partition key component k"'.
```

seems related to b7bbcdd178088c9fdfd217e9a419821d3a87c345 which isn't in the pushed branch.

I don't understand how test-timestamp.py passes in master after e1513f11991c2cd8b378132419472525fdcbf24e

bhalevy commented:
@nyh I tried running ./test.py --mode=dev cql-pytest in master and it didn't run `test-timestamp.py` at all.
I suspect that pytest ignores it since it doesn't use the `test_` naming convention.
(and `./test.py --mode=$mode test-timestamp` fails with: `Test test-timestamp not found`)

With `git mv test/cql-pytest/test{-,_}timestamp.py` it is found and fails as in the above CI.
I'll submit a PR to fix and rename the test.

nyh commented:
Not running one of the tesfs is a test.py bug, that thinks it knows better than pytest what to run. If you run test/cql-pytest/run, it will run pytest and it will also run this test.
In any case, last week this test file was renamed (from dash to underscore) so you won't have this problem any more.

nyh commented:
Not running one of the tesfs is a test.py bug, that thinks it knows better than pytest what to run. If you run test/cql-pytest/run, it will run pytest and it will also run this test.
In any case, last week this test file was renamed (from dash to underscore) so you won't have this problem any more.

To clarify:
1. pytest looks at files test*, not test_* as test files. Using the pattern test_* is a test.py bug.
2. This test file was already reamed when we noticed its odd name last week, you don't need to send another patch to do that.

bhalevy commented:
> 2. This test file was already reamed when we noticed its odd name last week, you don't need to send another patch to do that.

@nyh I don't see such PR nor a patch on the mailing list.
Care to provide a reference so I know when to rebase and on what?

bhalevy commented:
> @bhalevy this is so funny, we managed to confuse each other. It was **you**, who already added a patch in **this** PR, to rename test-timestamp.py to test_using_timestamp.py. It wasn't another PR - it was this one :-) So you don't need to submit another PR to rename this patch, you already did, this one...
> 
> Of course, if you want you can split out this rename from this PR and send it as a **separate** PR. Maybe that's what you meant you wanted to do, and I just misunderstood you?

Yes, I want to separate it since it's fixing a preexisting issue and would have different backport requirements.

bhalevy commented:
https://github.com/scylladb/scylladb/pull/14293

bhalevy commented:
[26ff8f7](https://github.com/scylladb/scylladb/commit/26ff8f7bf78a75f7e513e701e5f945ed6b311b9d): fixed warnings in docs/cql/dml.rst regarding empty lines in bullet-points list

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1939/

DoronArazii commented:
@avikivity please merge
